trunk/source/TraceAbstraction/src/de/uni_freiburg/informatik/ultimate/plugins/generator/traceabstraction/witnesschecking/WitnessUtils.java

```diff
@@ -34,50 +36,63 @@
-				witnessAutomaton, csToolkit, predicateFactory, stateFactoryForRefinement);
-
-		final LineCoverageCalculator<LETTER> origCoverage = new LineCoverageCalculator<>(services, abstraction);
-		final IDoubleDeckerAutomaton<LETTER, IPredicate> newAbstraction;
+	private static <LETTER extends IIcfgTransition<?>> IDoubleDeckerAutomaton<LETTER, IPredicate> reduce(
+			final INwaOutgoingLetterAndTransitionProvider<LETTER, IPredicate> automaton, final Property property,
+			final AutomataLibraryServices services) throws AutomataOperationCanceledException {
 		switch (property) {
 		case NON_REACHABILITY:
-			newAbstraction = new RemoveDeadEnds<>(new AutomataLibraryServices(services), wpa).getResult();
-			break;
+			return new RemoveDeadEnds<>(services, automaton).getResult();
 		case TERMINATION:
-			newAbstraction = new RemoveNonLiveStates<>(new AutomataLibraryServices(services), wpa).getResult();
-			break;
+			return new RemoveNonLiveStates<>(services, automaton).getResult();
 		default:
 			throw new AssertionError();
 		}
+	}
+
+	public static <LETTER extends IIcfgTransition<?>> IDoubleDeckerAutomaton<LETTER, IPredicate>
+			constructGraphMLWitnessProduct(final IUltimateServiceProvider services,
+					final INwaOutgoingLetterAndTransitionProvider<LETTER, IPredicate> abstraction,
+					final INwaOutgoingLetterAndTransitionProvider<WitnessEdge, WitnessNode> witnessAutomaton,
+					final PredicateFactory predicateFactory, final ILogger logger, final Property property)
+					throws AutomataOperationCanceledException {
+		final GraphMLWitnessProductAutomaton<LETTER> wpa =
+				new GraphMLWitnessProductAutomaton<>(services, abstraction, witnessAutomaton, predicateFactory);
+		final LineCoverageCalculator<LETTER> origCoverage = new LineCoverageCalculator<>(services, abstraction);
+		final IDoubleDeckerAutomaton<LETTER, IPredicate> newAbstraction =
+				reduce(wpa, property, new AutomataLibraryServices(services));
 		logger.info(wpa.generateBadWitnessInformation());
 		new LineCoverageCalculator<>(services, abstraction, origCoverage).reportCoverage("Witness product");
 		return newAbstraction;
 	}
 
+	public static <LETTER extends IIcfgTransition<?>> IDoubleDeckerAutomaton<LETTER, IPredicate>
+			constructYamlWitnessProduct(final IUltimateServiceProvider services,
+					final INwaOutgoingLetterAndTransitionProvider<LETTER, IPredicate> abstraction,
+					final Witness witness, final PredicateFactory predicateFactory, final ILogger logger,
+					final Property property) throws AutomataOperationCanceledException {
+		final var result = new YamlWitnessProductAutomaton<>(abstraction, witness, predicateFactory);
+		logger.info(
+				"Constructing product of automaton with %d states and violation witness of the following lengths: %s",
+				abstraction.size(), witness.getEntries().stream().map(x -> ((ViolationSequence) x).getSegments().size())
+						.collect(Collectors.toList()));

```

Comment:
There is some indentation issue. Is this the result of an unfavorable formatter setting?

Reply:
Yes, but this might have been fixed now, this was before the Java 21 migration, where we also slightly adapted the formatter. I will just keep this for now, we will run the formatter anyway on the `dev` branch.